ddtrace/internal/telemetry/telemetry_writer.py

```diff
@@ -0,0 +1,213 @@
+    _instance = None  # type: ClassVar[Optional[TelemetryWriter]]
+    _lock = forksafe.Lock()  # type: ClassVar[forksafe.ResetObject]
+    sequence = 0  # type: ClassVar[int]
+
+    def __init__(self, endpoint):
+        # type: (str) -> None
+        super(TelemetryWriter, self).__init__(interval=_get_interval_or_default())
+
+        # switch endpoint to agent endpoint
+        self.url = endpoint  # type: str
+        self.encoder = JSONEncoderV2()
+        self._events_queue = []  # type: List[TelemetryRequest]
+        self._integrations_queue = []  # type: List[Dict]
+
+    def _send_request(self, request):
+        # type: (TelemetryRequest) -> httplib.HTTPResponse
+        """
+        Sends a telemetry request to telemetry intake [this will be switched the agent]
+        """
+        conn = get_connection(self.url)
+        rb_json = self.encoder.encode(request["body"])
+        resp = None
+        with StopWatch() as sw:
+            try:
+                conn.request("POST", self.url, rb_json, request["headers"])
+                resp = get_connection_response(conn)
+
+                t = sw.elapsed()
+                if t >= self.interval:
+                    log_level = logging.WARNING
+                else:
+                    log_level = logging.DEBUG
+
+                log.log(log_level, "sent %d in %.5fs to %s. response: %s", len(rb_json), t, self.url, resp.status)
+            finally:
+                conn.close()
+        return resp
+
+    def flush_integrations_queue(self):
+        # type () -> List[Dict]
+        """Returns a list of all integrations queued by integration_event"""
+        with self._lock:
+            integrations = self._integrations_queue
+            self._integrations_queue = []
+        return integrations
+
+    def flush_events_queue(self):
+        # type () -> List[TelemetryRequest]
+        """Returns a list of all integrations queued by classmethods"""
+        with self._lock:
+            requests = self._events_queue
+            self._events_queue = []
+        return requests
+
+    def queued_events(self):
+        # type () -> List[TelemetryRequest]
+        return self._events_queue
+
+    def queued_integrations(self):
+        # type () -> List[Dict]
+        return self._integrations_queue
+
+    def periodic(self):
+        integrations = self.flush_integrations_queue()
+        if integrations:
+            integrations_request = app_integrations_changed_telemetry_request(integrations)
+            TelemetryWriter.add_event(integrations_request)
+
+        requests = self.flush_events_queue()
+        requests_failed = []  # type: List[TelemetryRequest]
+        for request in requests:
+            resp = self._send_request(request)
+            if resp.status >= 300:
+                requests_failed.append(request)
+
+        with self._lock:
+            # add failed requests back to the requests queue
+            self._events_queue.extend(requests_failed)
```

Comment:
will we retry these forever?

Reply:
For now yes. I'll fix this in a future commit

Reply:
This commit should solve this: 13678bd